src/rust/inetstack/protocols/layer4/tcp/established/sender.rs

```diff
@@ -393,68 +402,96 @@ impl Sender {
         cb.emit(header, Some(data));
     }
 
-    // This segment acknowledges new data, so process the ack.
-    pub fn process_ack(&mut self, mut cb: SharedControlBlock, header: &TcpHeader, now: Instant) {
-        let bytes_acknowledged: u32 = (header.ack_num - self.send_unacked.get()).into();
-        let mut bytes_remaining: usize = bytes_acknowledged as usize;
-        // Remove bytes from the unacked queue.
-        while bytes_remaining != 0 {
-            if let Some(mut segment) = self.unacked_queue.try_pop() {
-                // Add sample for RTO if we have an initial transmit time.
-                // Note that in the case of repacketization, an ack for the first byte is enough for the time sample.
-                // TODO: TCP timestamp support.
-                if let Some(initial_tx) = segment.initial_tx {
-                    cb.rto_add_sample(now - initial_tx);
-                }
+    // Process an ack.
+    pub fn process_ack(&mut self, header: &TcpHeader, now: Instant) {
+        // Process the ACK.
+        // Start by checking that the ACK acknowledges something new.
+        // TODO: Look into removing Watched types.
+        //
+        let send_unacknowledged: SeqNumber = self.send_unacked.get();
+        let send_next: SeqNumber = self.send_next.get();
 
-                if segment.bytes.len() > bytes_remaining {
-                    // Only some of the data in this segment has been acked.  Remove just the acked amount.
-                    expect_ok!(
-                        segment.bytes.adjust(bytes_remaining),
-                        "'segment' should contain at least 'bytes_remaining'"
-                    );
-                    segment.initial_tx = None;
-
-                    // Leave this segment on the unacknowledged queue.
-                    self.unacked_queue.push_front(segment);
-                    break;
-                }
+        if send_unacknowledged < header.ack_num {
+            // Remove the now acknowledged data from the unacknowledged queue, update the acked sequence number
+            // and update the sender window.
+
+            let bytes_acknowledged: u32 = (header.ack_num - self.send_unacked.get()).into();
```

Comment:
`bytes_acknowledged` could directly be of the type `usize` instead, to avoid typecasting it on the next line.

Reply:
This needs to be a u32 because we can only convert SeqNumber into u32